builtin/v9/migration/verifreg.go

```diff
@@ -0,0 +1,143 @@
+package migration
+
+import (
+	"context"
+
+	verifreg8 "github.com/filecoin-project/go-state-types/builtin/v8/verifreg"
+
+	"github.com/filecoin-project/go-address"
+	"github.com/filecoin-project/go-state-types/abi"
+	"github.com/filecoin-project/go-state-types/builtin"
+	market8 "github.com/filecoin-project/go-state-types/builtin/v8/market"
+	adt8 "github.com/filecoin-project/go-state-types/builtin/v8/util/adt"
+	init9 "github.com/filecoin-project/go-state-types/builtin/v9/init"
+	market9 "github.com/filecoin-project/go-state-types/builtin/v9/market"
+	adt9 "github.com/filecoin-project/go-state-types/builtin/v9/util/adt"
+	verifreg9 "github.com/filecoin-project/go-state-types/builtin/v9/verifreg"
+	"github.com/ipfs/go-cid"
+	"golang.org/x/xerrors"
+)
+
+func migrateVerifreg(ctx context.Context, adtStore adt8.Store, priorEpoch abi.ChainEpoch, initStateV9 init9.State, marketStateV8 market8.State, verifregStateV8 verifreg8.State, emptyMapCid cid.Cid) (cid.Cid, map[abi.DealID]verifreg9.AllocationId, error) {
+	pendingProposals, err := adt8.AsSet(adtStore, marketStateV8.PendingProposals, builtin.DefaultHamtBitwidth)
+	if err != nil {
+		return cid.Undef, nil, xerrors.Errorf("failed to load pending proposals: %w", err)
+	}
+
+	proposals, err := market8.AsDealProposalArray(adtStore, marketStateV8.Proposals)
+	if err != nil {
+		return cid.Undef, nil, xerrors.Errorf("failed to get proposals: %w", err)
+	}
+
+	pendingMap := make(map[abi.DealID]market8.DealProposal)
+	var proposal market8.DealProposal
+	if err = proposals.ForEach(&proposal, func(dealID int64) error {
```

Comment:
This is a large map (every deal currently active). It would be much faster to iterate PendingProposals and look them up in Proposals.

Reply:
Oh right, I see. There's good reason for the set of CIDs.